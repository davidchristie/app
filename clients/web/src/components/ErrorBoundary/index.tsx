import { Component, ErrorInfo, ReactNode } from "react";

export interface ErrorBoundaryProps {
  children: ReactNode;
  onError?: (error: Error, errorInfo: ErrorInfo) => void;
}

interface ErrorBoundaryState {
  error: Error | null;
}

export class ErrorBoundary extends Component<
  ErrorBoundaryProps,
  ErrorBoundaryState
> {
  public static getDerivedStateFromError(error: Error): ErrorBoundaryState {
    return { error };
  }

  public state: ErrorBoundaryState = {
    error: null,
  };

  public componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    if (this.props.onError) {
      this.props.onError(error, errorInfo);
    }
  }

  public render() {
    if (this.state.error !== null) {
      return <h1>{this.state.error.message}</h1>;
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
