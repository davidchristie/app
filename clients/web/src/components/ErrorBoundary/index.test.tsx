import { render, screen } from "@testing-library/react";
import ErrorBoundary from ".";

describe("ErrorBoundary", () => {
  describe("when no error is thrown", () => {
    it("renders correctly", () => {
      function Child() {
        return <div data-testid="Child" />;
      }
      const { container } = render(
        <ErrorBoundary>
          <Child />
        </ErrorBoundary>
      );
      const child = screen.getByTestId("Child");
      expect(child).toBeDefined();
      expect(child).toBe(container.firstChild);
    });
  });

  describe("when error is thrown", () => {
    const message = "Something went wrong";
    const error = new Error(message);

    let consoleError: jest.SpyInstance;

    function ThrowError(): JSX.Element {
      throw error;
    }

    beforeEach(() => {
      consoleError = jest.spyOn(console, "error");
      consoleError.mockImplementation(() => {});
    });

    afterEach(() => {
      consoleError.mockRestore();
    });

    it("renders correctly", () => {
      const { container } = render(
        <ErrorBoundary>
          <ThrowError />
        </ErrorBoundary>
      );
      expect(screen.getByText(message)).toBeDefined();
      expect(container.firstChild).toMatchSnapshot();
    });

    it("calls onError handler", () => {
      const onError = jest.fn();
      render(
        <ErrorBoundary onError={onError}>
          <ThrowError />
        </ErrorBoundary>
      );
      expect(onError).toBeCalledTimes(1);
      expect(onError).toBeCalledWith(error, expect.any(Object));
    });

    it("logs error to console", () => {
      render(
        <ErrorBoundary>
          <ThrowError />
        </ErrorBoundary>
      );
      expect(consoleError).toBeCalledTimes(2);
      expect(consoleError.mock.calls[0][0]).toContain(
        `Error: Uncaught [Error: ${message}]`
      );
    });
  });
});
