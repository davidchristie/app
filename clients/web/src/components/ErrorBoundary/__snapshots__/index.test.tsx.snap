// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary when error is thrown renders correctly 1`] = `
<h1>
  Something went wrong
</h1>
`;
